File: src/main/java/com/dotgoing/utils/option/None.java
package com.dotgoing.utils.option;

import java.util.function.Function;

public final class None<T> extends Option<T> {

    private final Exception defaultError = new Exception("none default error");
    private final Exception err;

    None() {
        err = defaultError;
    }

    None(Exception err) {
        this.err = err;
    }

    @Override
    public Exception error() {
        return err;
    }

    @Override
    public <R> Option<R> map(Function<? super T, ? extends R> transformer) {
        return Option.empty();
    }

    @Override
    public <R> Option<R> flatMap(Function<? super T, ? extends Option<? extends R>> transformer) {
        return Option.empty();
    }

    @Override
    public T value() {
        throw new RuntimeException("none has no value");
    }

    @Override
    public boolean hasValue() {
        return false;
    }

    @Override
    public String toString() {
        return "None(" + err.toString() + ")";
    }
}

File: src/main/java/com/dotgoing/utils/core/None.kt
package com.dotgoing.utils.core

class None<out T>(val exception: OptionException = OptionException()) : Option<T>() {
    override fun hasValue(): Boolean {
        return false
    }

    override fun value(): T {
        throw OptionStateException("none obj has no value : ${this}")
    }

    override fun errorCode(): Int {
        return exception.code
    }

    override fun error(): OptionException {
        return exception
    }

    override fun toString(): String {
        return exception.message ?: "none error"
    }
}